src/coreclr/jit/lower.cpp

```diff
@@ -3165,8 +3166,53 @@ void Lowering::LowerStoreLocCommon(GenTreeLclVarCommon* lclStore)
                 return;
             }
 #endif // !WINDOWS_AMD64_ABI
+            convertToStoreObj = false;
         }
-        else if (!src->OperIs(GT_LCL_VAR) || (varDsc->GetLayout()->GetRegisterType() == TYP_UNDEF))
+        else if (!varDsc->IsEnregisterable())
```

Comment:
the main goal of this change is to keep `ASG(LCL_VAR, 0)` as `STORE_LCL_VAR(0)` without taking the address of the lcl.